libp2p/muxers/yamux/yamux.nim

```diff
@@ -265,7 +265,9 @@ method readOnce*(
     raise newLPStreamRemoteClosedError()
   if channel.recvQueue.len == 0:
     channel.receivedData.clear()
-    await channel.closedRemotely or channel.receivedData.wait()
+    try:  # https://github.com/status-im/nim-chronos/issues/516
+      discard await race(channel.closedRemotely, channel.receivedData.wait())
+    except ValueError: raiseAssert("Futures list is not empty")
```

Comment:
You mean, for `race`? That's a chronos function, and agree that a defect would make this less annoying to call.

Or, you mean, when `ValueError` happens, to raise `Defect`? That's what `raiseAssert` does.

Reply:
I mean the second.  I guess in this proc we are reading from one particular stream. Do we want to crash the whole application in case of sth unexpected?

Reply:
`ValueError` is raised only if `race()` is called _without_ any arguments. There is no other way how `race` can raise a `ValueError`. This means that here, `race()` will strictly never raise a `ValueError`, that's what's asserted here.

The individual `closedRemotely` and `receivedData.wait` futures are not `await`-ed (neither are they in the old design, where the `or` operation is `await`-ed). Therefore, any errors from those are suppressed and don't bubble up here.

Reply:
What you described is the current behaviour, but this code might and most likely will change over time and ValueError could be raised. I believe the point is, what is the benefit of raising a `Defect` when reading from a stream? if the objective is to avoid declaring errors in the `readOnce` proc, wouldn't it be better to just log the error? But probably we should raise a MuxerError or sth like that.

Reply:
the `ValueError` has been removed in chronos master for the relevant `race` function

Reply:
@arnetheduck can you tag a new chronos release? libp2p testing uses nimble.

Reply:
The `Defect` is not handling 'reading from a stream' or 'readOnce', it is handling the error from `race`. The inner futures are not awaited and their errors are ignored.

Reply:
What I meant is that we are effectively raising a `Defect` cause of an error in `readOnce`, aren't we? It seems to me that regardless of what happens in this proc, it's not a reason to crash the system, unless I'm missing sth.

Reply:
`race` is annotated with `{.raises: [ValueError].}`. That `ValueError` is only raised if `race` is used incorrectly, as in, with an empty list of futures. That is not the case here, so it is never raised. This assumption is reflected in the `raiseAssert`.

The inner `channel.closedRemotely`, `channel.receivedData.wait` futures are _not_ `await`-ed, their errors are not raised unless doing a second `await` on them, or trying to access their results. The code does neither of that. `race` only waits for one of them to finish (either successfully, or error, or cancel), but doesn't care about the result and also doesn't re-raise potential errors.

Reply:
> tag a new chronos release

yeah, it's coming

Reply:
@etan-status we had this PR https://github.com/vacp2p/nim-libp2p/pull/902  to remove `if isSome() / get()` pattern from the codebase to avoid potential sources of `Defects`, even though, in theory, it wouldn't happen in the previous code. It seems counterproductive to me to explicitly raise `Defects` now.

Reply:
What's your preference for handling unreachable code?

Reply:
In this specific case, how about closing the channel and logging a warning?